pkg/clusterresource/controller.go

```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
 	return k8sManifest, nil
 }
 
+// createPatch create 3-way merge patch of current object, original object (retrieved from last applied annotation), and the modification
+// for native k8s resource, strategic merge patch is used
+// for custom resource, json merge patch is used
+// heavily inspired by kubectl's patcher
+func (c *controller) createPatch(mapping *meta.RESTMapping, currentObj *unstructured.Unstructured, modified []byte, namespace string) ([]byte, types.PatchType, error) {
+	current, err := k8sruntime.Encode(unstructured.UnstructuredJSONScheme, currentObj)
+	if err != nil {
+		return nil, "", fmt.Errorf("Failed to encode [%+v] in namespace [%s] to json with err: %v",
+			currentObj.GetKind(), namespace, err)
+	}
+
+	original, err := getLastApplied(currentObj)
+
+	var patch []byte
+	patchType := types.StrategicMergePatchType
+	obj, err := scheme.Scheme.New(mapping.GroupVersionKind)
+	switch {
+	case err == nil:
+		patchType = types.StrategicMergePatchType
+		if patch == nil {
+			lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(obj)
+			if err != nil {
+				return nil, "", fmt.Errorf("Failed to create lookup patch meta for [%+v] in namespace [%s] with err: %v",
+					currentObj.GetKind(), namespace, err)
+			}
+
+			patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, true)
+			if err != nil {
+				return nil, "", fmt.Errorf("Failed to create 3 way merge patch for resource [%+v] in namespace [%s] with err: %v\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s",
+					currentObj.GetKind(), namespace, err, original, modified, current)
+			}
+		}
+	case k8sruntime.IsNotRegisteredError(err):
```

Comment:
what does not registered in this case mean?

Reply:
It's resources that's not recognized by client-go's schema by default. It's usually CRDs